docs/design/README.md

```diff
@@ -0,0 +1,552 @@
+name = "twoliter"
+version = "v0.1.0"
+```
+
+### Directory Structure
+
+Project directory structure is similar to what we see in Bottlerocket's main repo:
+
+```
+.
+├── build
+├── kits
+│   └── hello-dev-kit
+│       ├── src
+│       │   ├── kit.rs
+│       │   └── kit.spec
+│       └── Cargo.toml
+├── packages
+│   └── hello-agent
+│       ├── build.rs
+│       ├── Cargo.toml
+│       ├── hello-agent.service
+│       ├── hello-agent.spec
+│       └── pkg.rs
+├── sources
+│   ├── hello-agent
+│   │   ├── src
+│   │   │   └── main.rs
+│   │   └── Cargo.toml
+│   ├── Cargo.toml
+│   ├── clarify.toml
+│   └── deny.toml
+├── variants
+│   ├── example-dev
+│   │   ├── build.rs
+│   │   ├── Cargo.toml
+│   │   ├── pkg.rs
+│   │   └── variant.spec
+│   └── Cargo.toml
+├── COPYRIGHT
+├── LICENSE-APACHE
+└── LICENSE-MIT
+```
+
+Directory descriptions:
+
+* build: The output directory for Bottlerocket images, RPM files, etc.
+* kits: Where the local kits are defined.
+* packages: Where the RPM packages are defined.
+* variants: Where Bottlerocket image builds are defined.
+
+### Twoliter New
+
+Inspired by `cargo new`, we will implement a `twoliter new` command that creates a new Twoliter project.
+The command creates a directory structure like the one shown above.
+The new project will include a minimal variant with a simple custom package and settings.
+The command will take a few arguments, such as `--project name` and `--variant-name`.
+
+Similarly, a command will be available to add a new variant to an existing project.
+For example, `twoliter new-variant` which will add a variant.
+
+```
+❯ twoliter new project --name ootb-variants --variant-name example-dev
+❯ cd ootb-variants
+❯ twoliter build --variant hello-dev
+  ...
+❯ twoliter new-variant my-new-variant
+```
+
+### Twoliter Update
+
+`twoliter update` is another inspiration from `cargo`.
+This command makes it easy for maintainers to update the packages (kits actually) they are using to the latest versions.
+
+For simplicity, no two variants in a project can use differing versions of a single kit.
+Twoliter will have the ability to create a kit dependency tree with the following procedure.
+First Twoliter will read the kit dependencies from each variant and add any external kits to its graph.
+Next Twoliter will read the kit definitions in the `kits` directory, and add all of their external dependencies to its graph.
+
+Once it has this list, Twoliter will begin a traversal.
+For each kit it will pull the `metdata` container for each external kit and store the kit.json files locally.
+Each time it finds a new external kit (i.e. an external kit that is depended on by an external kit), it will add this new dependency to the graph.
+(TODO: are cycles OK? They might be.)
+
+At this point Twoliter will determine the maximum version of each kit that is possible without introducing multiple versions of the same kit.
+TODO: this is probably hard, but there is prior art to look at.
```

Comment:
You may actually need to store things as some sort of reference tree, since the dependency closure can change based on which minor version of a package you select. Tricky.

Reply:
Can we take advantage of an existing package manager to do this somehow?